compaction/compaction_manager.cc

```diff
@@ -1037,8 +1037,11 @@ future<> compaction_manager::remove(column_family* cf) {
         // Wait for all functions running under with_compaction_disabled_gate to terminate.
         auto close_gate = csptr->with_compaction_disabled_gate.close();
 
+        // Acquire the write lock to make sure all tasks that acquired it are done.
+        auto write_lock = csptr->lock.write_lock();
+
```

Comment:
Can this cause a deadlock? Does it depend on the lock being FIFO?

Reply:
It can't cause a deadlock since the `compaction_state` is detached and the (undocumented) reference/locking order calls for first taking a shared_ptr reference on the compaction_state, via `this->_compaction_state`, then acquiring either read or write lock.

Once detach, acquiring the `compaction_state_ptr->lock.write_lock()` must to the last in line so it serializes after all present lock holders.

To answer your question, yes it depends on the lock being FIFO.

Note the the "gated lock" idea I submitted in seastar-dev is targeting this scenario so we can call `lock.close()` that will wait on all lock holders AND will prevent new locks to be acquired, without depending on the lock FIFO property.

Reply:
This looks very complicated. I'm sure there is a simpler solution.

Reply:
The simpler solution is entering the compaction_state gate for all functions referencing it. These will be compaction tasks of all kinds and functions running under `run_with_compaction_disabled`.

Reply:
I think we're making wrong usage of this lock. It's only used for serializing minor and major compactions. Today we have the gate to wait on tasks running under it, and the future associated with each task that can be waited upon. No need to acquire the write lock on the removal path. I think we can drop this change.

Reply:
this is dropped in v2.
instead the compaction_state::gate will cover all uses of compaction_state so that
the gate is held while `compaction_state` is being used.